lib/types.nix

```diff
@@ -608,7 +608,11 @@ rec {
               then lhs.specialArgs // rhs.specialArgs
               else throw "A submoduleWith option is declared multiple times with the same specialArgs \"${toString (attrNames intersecting)}\"";
             shorthandOnlyDefinesConfig =
-              if lhs.shorthandOnlyDefinesConfig == rhs.shorthandOnlyDefinesConfig
+              if lhs.shorthandOnlyDefinesConfig == null
+              then rhs.shorthandOnlyDefinesConfig
+              else if rhs.shorthandOnlyDefinesConfig == null
+              then lhs.shorthandOnlyDefinesConfig
+              else if lhs.shorthandOnlyDefinesConfig == rhs.shorthandOnlyDefinesConfig
```

Comment:
Since `null` is now an allowed value, we should also define its semantics and document it. The semantics should be that if this value is needed to make a decision and it's `null`, an error is thrown. This means that if you do any shorthand definitions, you should get an error, unless another option declaration specifies the value of `shorthandOnlyDefinesConfig`, e.g. whereas

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith {
          modules = [];
          shorthandOnlyDefinesConfig = null;
        };
      };

      config.foo.bar = 10;
    }
  ];
}
```

with this PR currently throws
```
error: value is null while a Boolean was expected

       at /home/infinisil/src/nixpkgs/lib/types.nix:538:23:

          537|           then setFunctionArgs (args: unify (value args)) (functionArgs value)
          538|           else unify (if shorthandOnlyDefinesConfig then { config = value; } else value);
             |                       ^
          539|
(use '--show-trace' to show detailed location information)
```

it should instead be something like

```
error: shorthandOnlyDefinesConfig is null
(use '--show-trace' to show detailed location information)
```

Reply:
> Since `null` is now an allowed value

I disagree. You'd have to read the implementation to find that null is acceptable in some situations.

To keep things simple and, I've decided to default it to `true`. I don't see much point in documenting this though.